### PR TITLE
Add KMeterISO component.

### LINKED ISSUE #2333: Support for M5Stack KMeterISO
**Describe the problem you have/What new integration you would like**

The KMeterISO device is an isolated thermometer based around a K-type thermocouple which communicates through I2C on address 0x66.

https://docs.m5stack.com/en/unit/KMeterISO%20Unit contains documentation on the device.

A simple library that directly uses I2C (does not use the I2C abstraction shipped with ESPHome) is available here: https://github.com/m5stack/M5Unit-KMeterISO .

Full code for the working component which I just finished is at https://github.com/Rudd-O/kmeteriso .  I have made a PR for it (linked below in comments) which you are free to release it under whatever license you want.  I did not know in the end how to select which I2C bus (of the chips that support multiple buses) was to be used, so it always uses the primary bus as far as I know.

**Please describe your use case for this integration and alternatives you've tried:**

This would enable customers that have bought this device to use it.  Right now I have no way to use the KMeterISO probe using ESPHome.

**Additional context**

No context.

### PR DESCRIPTION
**Pull request by Rudd-O**
# What does this implement/fix?

KMeterISO K-Type thermocouple support, pursuant to https://github.com/esphome/feature-requests/issues/2333 .

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/2333

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3095

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
i2c:
  sda: GPIO21
  scl: GPIO22
  scan: true
  id: bus_a
  frequency: 100kHz

sensor:
- platform: kmeteriso
  temperature:
    name: Temperature
  internal_temperature:
    name: Internal temperature
    entity_category: diagnostic
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by Rudd-O** [comment]
~Folks, I need help with the tests and the clang format.  Not sure what is going on.  Please help me either with pointers or do feel free to add commits to the pull request.  Thanks so much in advance!~

This is ready for review.

File: esphome/components/kmeteriso/kmeteriso.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
+#include "kmeteriso.h"
+#include "esphome/core/hal.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace kmeteriso {
+
+static const char *const TAG = "kmeteriso.sensor";
+
+static const uint8_t KMETER_KMETER_ERROR_STATUS_REG = 0x20;
```
Comment (jesserockz) [inline esphome/components/kmeteriso/kmeteriso.cpp]:
```suggestion
static const uint8_t KMETER_ERROR_STATUS_REG = 0x20;
```

File: esphome/components/kmeteriso/kmeteriso.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+float KMeterISOComponent::get_setup_priority() const { return setup_priority::DATA; }
+
+void KMeterISOComponent::update() {
+  uint8_t read_buf[4];
+
+  if (!this->read_bytes(KMETER_TEMP_VAL_REG, read_buf, 4)) {
+    ESP_LOGW(TAG, "Error reading temperature.");
+  } else {
+    int32_t temp = (read_buf[3] << 24) | (read_buf[2] << 16) | (read_buf[1] << 8) | read_buf[0];
```
Comment (jesserockz) [inline esphome/components/kmeteriso/kmeteriso.cpp]:
```suggestion
    int32_t temp = encode_uint32(read_buf[3], read_buf[2], read_buf[1], read_buf[0]);
```

File: esphome/components/kmeteriso/kmeteriso.cpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    ESP_LOGV(TAG, "Got temperature=%.2f °C", temp_f);
+    if (this->temperature_sensor_ != nullptr)
+      this->temperature_sensor_->publish_state(temp_f);
+  }
+
+  if (!this->read_bytes(KMETER_INTERNAL_TEMP_VAL_REG, read_buf, 4)) {
+    ESP_LOGW(TAG, "Error reading internal temperature.");
+    return;
+  } else {
+    int32_t internal_temp = (read_buf[3] << 24) | (read_buf[2] << 16) | (read_buf[1] << 8) | read_buf[0];
```
Comment (jesserockz) [inline esphome/components/kmeteriso/kmeteriso.cpp]:
```suggestion
    int32_t internal_temp = encode_uint32(read_buf[3], read_buf[2], read_buf[1], read_buf[0]);
```

File: esphome/components/kmeteriso/kmeteriso.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+void KMeterISOComponent::update() {
+  uint8_t read_buf[4];
+
+  if (!this->read_bytes(KMETER_TEMP_VAL_REG, read_buf, 4)) {
+    ESP_LOGW(TAG, "Error reading temperature.");
+  } else {
+    int32_t temp = (read_buf[3] << 24) | (read_buf[2] << 16) | (read_buf[1] << 8) | read_buf[0];
+    float temp_f = temp / 100.0;
+    ESP_LOGV(TAG, "Got temperature=%.2f °C", temp_f);
+    if (this->temperature_sensor_ != nullptr)
```
Comment (jesserockz) [inline esphome/components/kmeteriso/kmeteriso.cpp]: This check should be before the `read_bytes` so it doesnt have to communicate if the sensor is not configured. Same for the below sensor.

File: esphome/components/kmeteriso/kmeteriso.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#include "esphome/components/i2c/i2c_bus.h"
+
+namespace esphome {
+namespace kmeteriso {
+
+/// This class implements support for the KMeterISO thermocouple sensor.
+class KMeterISOComponent : public PollingComponent, public i2c::I2CDevice {
+ public:
+  void set_temperature_sensor(sensor::Sensor *t) { temperature_sensor_ = t; }
+  void set_internal_temperature_sensor(sensor::Sensor *t) { internal_temperature_sensor_ = t; }
```
Comment (jesserockz) [inline esphome/components/kmeteriso/kmeteriso.h]:
```suggestion
  void set_temperature_sensor(sensor::Sensor *t) { this->temperature_sensor_ = t; }
  void set_internal_temperature_sensor(sensor::Sensor *t) { this->internal_temperature_sensor_ = t; }
```

File: esphome/components/kmeteriso/kmeteriso.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+  // ========== INTERNAL METHODS ==========
+  // (In most use cases you won't need these)
+  void setup() override;
+  float get_setup_priority() const override;
+  void update() override;
+
+ protected:
+  /// Read the temperature value and store the calculated ambient temperature in t_fine.
+  float read_temperature_(const uint8_t *data, int32_t *t_fine);
```
Comment (jesserockz) [inline esphome/components/kmeteriso/kmeteriso.h]: Not used?
```suggestion
```

File: esphome/components/kmeteriso/sensor.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                unit_of_measurement=UNIT_CELSIUS,
+                device_class=DEVICE_CLASS_TEMPERATURE,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+        }
+    )
+    .extend(cv.polling_component_schema("5s"))
+    .extend(i2c.i2c_device_schema(0x66))
+    # Unsure how the polling frequency could be dynamically
+    # changed by the end user of this component.
```
Comment (jesserockz) [inline esphome/components/kmeteriso/sensor.py]: Users can set `update_interval: XXs` in yaml to change the timer (as you added in the test yaml)

We also prefer to default everything to 60s

```suggestion
    .extend(cv.polling_component_schema("60s"))
    .extend(i2c.i2c_device_schema(0x66))
```

File: esphome/components/kmeteriso/sensor.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_TEMPERATURE in config:
+        conf = config[CONF_TEMPERATURE]
+        sens = await sensor.new_sensor(conf)
```
Comment (jesserockz) [inline esphome/components/kmeteriso/sensor.py]:
```suggestion
    if temperature_config := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_config)
```

File: esphome/components/kmeteriso/sensor.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        cg.add(var.set_temperature_sensor(sens))
+    if CONF_INTERNAL_TEMPERATURE in config:
+        conf = config[CONF_INTERNAL_TEMPERATURE]
+        sens = await sensor.new_sensor(conf)
```
Comment (jesserockz) [inline esphome/components/kmeteriso/sensor.py]:
```suggestion
    if internal_temperature_config := config.get(CONF_INTERNAL_TEMPERATURE):
        sens = await sensor.new_sensor(internal_temperature_config)
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/kmeteriso/kmeteriso.cpp (hunk 2 above)
Comment (Rudd-O) [reply esphome/components/kmeteriso/kmeteriso.cpp]: Wow.  Great knowledge and nice trick!  Thanks.

**Comment by Rudd-O** [comment]
Thanks for the review.  All review suggestions were accepted.  PTAL @jesserockz .

**Comment by Rudd-O** [comment]
image

Strange.  No review comment was unaddressed.

**Review by jesserockz — DISMISSED** [review]

**Comment by jesserockz** [comment]
Thanks for making the changes

**Comment by Rudd-O** [comment]
Despite the approval, I can't merge this — or maybe the UI will allow me to merge once the tests finish.  Will wait a little longer.

**Comment by jesserockz** [comment]
No, you cannot merge, only people with write access can merge (like myself) which is done when both code and docs are approved

**Comment by Rudd-O** [comment]
> No, you cannot merge, only people with write access can merge (like myself) which is done when both code and docs are approved

Cool.  Thanks!

**Comment by Rudd-O** [comment]
Check your Discord privmsgs please @jesserockz .

**Review by jesserockz — APPROVED** [review]

**Comment by Rudd-O** [comment]
Woohoo.  Thanks.

**Pull request merged**